File: hw/hw1/6-8.c
/*
 *   Circuit Satisfiability, Version 3
 *
 *   This version of the program prints the total number of
 *   solutions and the execution time.
 *
 *   Programmed by Michael J. Quinn
 *
 *   Last modification: 3 September 2002
 */

#include "mpi.h"
#include <assert.h>
#include <stdio.h>

#define DATATYPE MPI_DOUBLE

int main(int argc, char* argv[])
{
    double elapsed_time; /* Time to find, count solutions */
    int id; /* Process rank */
    int p; /* Number of processes */
    MPI_Status status;

    MPI_Init(&argc, &argv);

    MPI_Comm_rank(MPI_COMM_WORLD, &id);
    MPI_Comm_size(MPI_COMM_WORLD, &p);
    MPI_Barrier(MPI_COMM_WORLD);
    if (id == 0) {
        printf("Processer number %d\n\n", p);
        fflush(stdout);
    }
    // Start Here
    assert(p == 2);

    /* Start timer */
    elapsed_time = -MPI_Wtime();

    int t = 1e2;
    int DUMMY_TAG = 0;
    double sent;
    double received;
    int size = sizeof(sent) / 4;

    for (int i = 0; i < t; i++) {
        sent = i;
        if (id == 0) {
            int destId = 1;
            
            MPI_Send(&sent, size, DATATYPE, destId, DUMMY_TAG, MPI_COMM_WORLD);
            MPI_Recv(&received, size, DATATYPE, destId, DUMMY_TAG, MPI_COMM_WORLD, &status);
        }
        if (id == 1) {
            int sourceId = 0;

            MPI_Recv(&received, size, DATATYPE, sourceId, DUMMY_TAG, MPI_COMM_WORLD, &status);
            MPI_Send(&received, size, DATATYPE, sourceId, DUMMY_TAG, MPI_COMM_WORLD);
        }
    }
    /* Stop timer */
    elapsed_time += MPI_Wtime();

    // meta
    if (id == 0) {
        printf("t = %d\n", t);
        fflush(stdout);
        printf("n-byte, n = %d\n", sizeof(sent));
        fflush(stdout);
        printf("Execution time %8.6f\n", elapsed_time);
        fflush(stdout);
        printf("Avg latency + (n / bandwith) = %8.10f\n", elapsed_time / (2 * t));
        fflush(stdout);
    }

    MPI_Finalize();

    return 0;
}

File: hw/hw1/6-8.py
# t = 100000
# n-byte, n = 8
# Execution time 0.142340
# Avg latency + (n / bandwith) = 0.0000007117

# t = 100000
# n-byte, n = 4
# Execution time 0.135645
# Avg latency + (n / bandwith) = 0.0000006782

# latency + (8 / bandwith) = 0.0000007117
# latency + (4 / bandwith) = 0.0000006782

time1 = 0.0000007117
n1 = 8
time2 = 0.0000006782
n2 = 4
time3 = 0.0000006040
n3 = 4

# time1 - time2 = (n1+n2)/b

def calLB(time1, n1, time2, n2):
    b = (n1+n2) / (time1-time2)
    l = time1 - (n1/b)

    return l, b
    
l, b = calLB(time1, n1, time2, n2)
print(l, b)

b = (n2+n3) / (time2 - time3)
l = time2 - (n2 / b)
print(l, b)

# cpu1
l, b = calLB(0.0000007299, 8, 0.0000007752, 4)
print(l, b)
# b = (n1+n2) / (time1 - time2)
